### PR TITLE
617 enhancement search keyword feature

### PR DESCRIPTION
**Pull request by itiB**
## What Changed

- Changed 1
- Changed 2

## Evidence

Please describe the functionality gained by merging this pull-request and the evidence that the bug has been fixed.

### CODE REVIEW

**Comment by hitenkoku** [comment]
I have confirmed that the existing test is no longer passing.

Caused by https://github.com/Yamato-Security/hayabusa/pull/893/files#diff-49ef28116581dce0b79a9d5d837aeaa51600d6914ff192deba071fb4292a4f In 02R220-R230, the string obtained for object used to be None, but now it is a value.

I'll go through the other tests and see if I can't make sure there are no problems.

**Comment by YamatoSecurity** [comment]
@hitenkoku Thank you!
when a user just puts in the `search` subcommand it gives:
Screen Shot 2023-04-07 at 8 34 28 PM
but can you have it display the same as `help search`:
Screen Shot 2023-04-07 at 8 34 38 PM

**Comment by YamatoSecurity** [comment]
@hitenkoku 
When using `search`, it displays the following:
```
Searching...

Start time: 2023/04/07 20:43

Total event log files: 583
Total file size: 137.1 MB

Excluded rules: 20
Noisy rules: 7 (Disabled)

Deprecated rules: 152 (4.38%) (Disabled)
Experimental rules: 1776 (51.12%)
Stable rules: 223 (6.42%)
Test rules: 1475 (42.46%)
Unsupported rules: 43 (1.24%) (Disabled)

Hayabusa rules: 149
Sigma rules: 3325
Total enabled detection rules: 3474

Scanning in progress. Please wait.
```

but as rules are not being used I would like to change to the following:

```
Start time: 2023/04/07 20:43

Total event log files: 583
Total file size: 137.1 MB

Search in progress. Please wait.
```

**Comment by YamatoSecurity** [comment]
When displaying to screen, fields are comma delimited and fields are delimited by `|`:
`2021-06-13T06:17:18.044373Z,sv-dc.hinokabegakure-no-sato.local,microsoft-windows-bits-client/operational,3,12,-,jobId:6F364A04-6299-4349-A8B6-34BED5E400EF | jobOwner:hns\\Administrator | jobTitle:test | processId:1020 | processPath:C:\\Windows\\System32\\bitsadmin.exe,../hayabusa-sample-evtx/YamatoSecurity/Persistence/T1197_BitsClient.evtx`

Is it possible to delimit them like in `csv-timeline` with `‖` and `¦` and put spaces between fields and their data?
Also, would like to abbreviate the channel name like in `csv-timeline`. 
Ex: `microsoft-windows-bits-client/operational` -> `BitsCli`

Ex:
`2021-06-13T06:17:18.044373Z ‖ sv-dc.hinokabegakure-no-sato.local ‖ BitsCli ‖ 3 ‖ 12 ‖ - ‖ jobId: 6F364A04-6299-4349-A8B6-34BED5E400EF ¦ jobOwner: hns\\Administrator ¦ jobTitle: test ¦ processId: 1020 ¦ processPath: C:\\Windows\\System32\\bitsadmin.exe ‖ ../hayabusa-sample-evtx/YamatoSecurity/Persistence/T1197_BitsClient.evtx`

**Comment by hitenkoku** [comment]
resolved following problem by https://github.com/Yamato-Security/hayabusa/pull/893/commits/becc21fb068e574a3c9e4dbdcb8128b09aaafa33

> I have confirmed that the existing test is no longer passing.
> 
> Caused by https://github.com/Yamato-Security/hayabusa/pull/893/files#diff-49ef28116581dce0b79a9d5d837aeaa51600d6914ff192deba071fb4292a4f In 02R220-R230, the string obtained for object used to be None, but now it is a value.
> 
> While passing other tests, I will make modifications so as not to impair functionality.

**Comment by hitenkoku** [comment]
resolved following remarks in https://github.com/Yamato-Security/hayabusa/pull/893/commits/c1d5db44ead4295522f7a63e56c17371061658bd

> When displaying to screen, fields are comma delimited and fields are delimited by `|`: `2021-06-13T06:17:18.044373Z,sv-dc.hinokabegakure-no-sato.local,microsoft-windows-bits-client/operational,3,12,-,jobId:6F364A04-6299-4349-A8B6-34BED5E400EF | jobOwner:hns\\Administrator | jobTitle:test | processId:1020 | processPath:C:\\Windows\\System32\\bitsadmin.exe,../hayabusa-sample-evtx/YamatoSecurity/Persistence/T1197_BitsClient.evtx`
> 
> Is it possible to delimit them like in `csv-timeline` with `‖` and `¦` and put spaces between fields and their data? Also, would like to abbreviate the channel name like in `csv-timeline`. Ex: `microsoft-windows-bits-client/operational` -> `BitsCli`
> 
> Ex: `2021-06-13T06:17:18.044373Z ‖ sv-dc.hinokabegakure-no-sato.local ‖ BitsCli ‖ 3 ‖ 12 ‖ - ‖ jobId: 6F364A04-6299-4349-A8B6-34BED5E400EF ¦ jobOwner: hns\\Administrator ¦ jobTitle: test ¦ processId: 1020 ¦ processPath: C:\\Windows\\System32\\bitsadmin.exe ‖ ../hayabusa-sample-evtx/YamatoSecurity/Persistence/T1197_BitsClient.evtx`

**Comment by hitenkoku** [comment]
resolved following remarks in https://github.com/Yamato-Security/hayabusa/pull/893/commits/a6ce3466cdfc764e95d58a09f3e3df6fddc15c36

> @hitenkoku When using `search`, it displays the following:
> 
> ```
> Searching...
> 
> Start time: 2023/04/07 20:43
> 
> Total event log files: 583
> Total file size: 137.1 MB
> 
> Excluded rules: 20
> Noisy rules: 7 (Disabled)
> 
> Deprecated rules: 152 (4.38%) (Disabled)
> Experimental rules: 1776 (51.12%)
> Stable rules: 223 (6.42%)
> Test rules: 1475 (42.46%)
> Unsupported rules: 43 (1.24%) (Disabled)
> 
> Hayabusa rules: 149
> Sigma rules: 3325
> Total enabled detection rules: 3474
> 
> Scanning in progress. Please wait.
> ```
> 
> but as rules are not being used I would like to change to the following:
> 
> ```
> Start time: 2023/04/07 20:43
> 
> Total event log files: 583
> Total file size: 137.1 MB
> 
> Search in progress. Please wait.
> ```

**Comment by hitenkoku** [comment]
@YamatoSecurity  
Added the feature to search for multiple keywords in https://github.com/Yamato-Security/hayabusa/pull/893/commits/0bb170fe79d46234c835f003dd5a527242bb0c0c .
```
PS >.\617r4.exe search -d ..\all-evtx\ -k IE -k powershell --debug -o 617r4.csv -q

...
Searching...

Start time: 2023/04/08 02:15

Total event log files: 1858
Total file size: 6.1 GB

1858 / 1858 [===============================================================================================] 100.00 %

Total findings: 60065


Elapsed time: 00:01:28.768
Rule Parse Processing Time: 00:00:01.519
Analysis Processing Time: 00:01:27.055
Output Processing Time: 00:00:00.183

Memory usage stats:
heap stats:    peak      total      freed    current       unit      count
  reserved:    1.5 GiB    6.0 GiB    4.5 GiB    1.5 GiB
 committed:  848.6 MiB   46.1 GiB   45.4 GiB  763.5 MiB
     reset:      0          0          0          0                            ok
   touched:   64.2 KiB   18.3 MiB   62.8 GiB  -62.7 GiB                        ok
  segments:     19        293        281         12                            not all freed!
-abandoned:      0          0          0          0                            ok
   -cached:      0          0          0          0                            ok
     pages:      0          0      752.2 Ki  -752.2 Ki                         ok
-abandoned:      0          0          0          0                            ok
 -extended:      0
 -noretire:      0
     mmaps:      0
   commits:   39.4 Ki
   threads:     32         32          0         32                            not all freed!
  searches:     0.0 avg
numa nodes:       1
   elapsed:      88.814 s
   process: user: 55.265 s, system: 1.984 s, faults: 2153935, rss: 640.6 MiB, commit: 701.3 MiB
```

**Comment by YamatoSecurity** [comment]
@hitenkoku Thanks for all the quick updates!
When I try to run with `-J` I get:
`[ERROR] --filepath only accepts .evtx files. Hidden files are ignored.`

Command: `./target/release/hayabusa search -J -f ./kosen-challenge.json -k sasuke`

JSON searching will be much different from evtx searching, no? Maybe we should remove the `-J` for now?
Or has it already be implemented? (or not too hard to implement?)

**Comment by YamatoSecurity** [comment]
> resolved following remarks in [c1d5db4](https://github.com/Yamato-Security/hayabusa/commit/c1d5db44ead4295522f7a63e56c17371061658bd)

Thank you! The headers now are separated by ` ‖ ` which looks nice but the fields are still separated by comma and do not have a space between them making it hard to read.
Ex:
```
 2022-05-24T04:59:06.909801Z,YamatoSecurity,KeyMgtSvc,1337,2,-,Binary:null ¦ Data:[FCE8820000006089E531C0648B50308B520C8B52148B72280FB74A2631FFAC3C617C022C20C1CF0D01C7E2F252578B52108B4A3C8B4C1178E34801D1518B592001D38B4918E33A498B348B01D631FFACC1CF0D01C738E075F6037DF83B7D2475E4588B582401D3668B0C4B8B581C01D38B048B01D0894424245B5B61595A51FFE05F5F5A8B12EB8D5D6A018D85B20000005068318B6F87FFD5BBF0B5A25668A695BD9DFFD53C067C0A80FBE07505BB4713726F6A0053FFD563616C632E65786500],../hayabusa-sample-evtx/YamatoSecurity/Persistence/SharpEventPersistHiddenShellcode.evtx,2021-06-13T06:17:18.087054Z,sv-dc.hinokabegakure-no-sato.local,BitsCli,59,14,-,Id:6F364A04-6299-4349-A8B6-34BED5E400EF ¦ bytesTotal:32256 ¦ bytesTransferred:0 ¦ bytesTransferredFromPeer:0 ¦ fileLength:32256 ¦ fileTime:2021-05-19T02:42:52.000000Z ¦ name:test ¦ peer: ¦ transferId:837FB303-0B01-460E-9024-FF2F735B3C82 ¦ url:http://192.168.10.254:80/calc.exe,../hayabusa-sample-evtx/YamatoSecurity/Persistence/T1197_BitsClient.evtx,2021-06-13T06:17:28.334195Z,sv-dc.hinokabegakure-no-sato.local,BitsCli,4,16,-,User:hns\\Administrator ¦ bytesTransferred:32256 ¦ bytesTransferredFromPeer:0 ¦ fileCount:1 ¦ jobId:6F364A04-6299-4349-A8B6-34BED5E400EF ¦ jobOwner:hns\\Administrator ¦ jobTitle:test,../hayabusa-sample-evtx/YamatoSecurity/Persistence/T1197_BitsClient.evtx,2021-06-13T06:17:18.044373Z,sv-dc.hinokabegakure-no-sato.local,BitsCli,3,12,-,jobId:6F364A04-6299-4349-A8B6-34BED5E400EF ¦ jobOwner:hns\\Administrator ¦ jobTitle:test ¦ processId:1020 ¦ processPath:C:\\Windows\\System32\\bitsadmin.exe,../hayabusa-sample-evtx/YamatoSecurity/Persistence/T1197_BitsClient.evtx,2021-06-13T06:17:18.056672Z,sv-dc.hinokabegakure-no-sato.local,BitsCli,310,13,-,ErrorCode:2147942450,../hayabusa-sample-evtx/YamatoSecurity/Persistence/T1197_BitsClient.evtx,
```

I would like to display like follows:
```
2022-05-24T04:59:06.909801Z ‖ YamatoSecurity ‖ KeyMgtSvc ‖ 1337 ‖ 2 ‖ - ‖ Binary:null ¦ Data:[FCE8820000006089E531C0648B50308B520C8B52148B72280FB74A2631FFAC3C617C022C20C1CF0D01C7E2F252578B52108B4A3C8B4C1178E34801D1518B592001D38B4918E33A498B348B01D631FFACC1CF0D01C738E075F6037DF83B7D2475E4588B582401D3668B0C4B8B581C01D38B048B01D0894424245B5B61595A51FFE05F5F5A8B12EB8D5D6A018D85B20000005068318B6F87FFD5BBF0B5A25668A695BD9DFFD53C067C0A80FBE07505BB4713726F6A0053FFD563616C632E65786500] ‖ ../hayabusa-sample-evtx/YamatoSecurity/Persistence/SharpEventPersistHiddenShellcode.evtx
2021-06-13T06:17:18.087054Z ‖ sv-dc.hinokabegakure-no-sato.local ‖ BitsCli ‖ 59 ‖ 14 ‖ - ‖ Id:6F364A04-6299-4349-A8B6-34BED5E400EF ¦ bytesTotal:32256 ¦ bytesTransferred:0 ¦ bytesTransferredFromPeer:0 ¦ fileLength:32256 ¦ fileTime:2021-05-19T02:42:52.000000Z ¦ name:test ¦ peer: ¦ transferId:837FB303-0B01-460E-9024-FF2F735B3C82 ¦ url:http://192.168.10.254:80/calc.exe ‖ ../hayabusa-sample-evtx/YamatoSecurity/Persistence/T1197_BitsClient.evtx
2021-06-13T06:17:28.334195Z ‖ sv-dc.hinokabegakure-no-sato.local ‖ BitsCli ‖ 4 ‖ 16 ‖ - ‖ User:hns\\Administrator ¦ bytesTransferred:32256 ¦ bytesTransferredFromPeer:0 ¦ fileCount:1 ¦ jobId:6F364A04-6299-4349-A8B6-34BED5E400EF ¦ jobOwner:hns\\Administrator ¦ jobTitle:test ‖ ../hayabusa-sample-evtx/YamatoSecurity/Persistence/T1197_BitsClient.evtx
2021-06-13T06:17:18.044373Z ‖ sv-dc.hinokabegakure-no-sato.local ‖ BitsCli ‖ 3 ‖ 12 ‖ - ‖ jobId:6F364A04-6299-4349-A8B6-34BED5E400EF ¦ jobOwner:hns\\Administrator ¦ jobTitle:test ¦ processId:1020 ¦ processPath:C:\\Windows\\System32\\bitsadmin.exe ‖ ../hayabusa-sample-evtx/YamatoSecurity/Persistence/T1197_BitsClient.evtx
2021-06-13T06:17:18.056672Z ‖ sv-dc.hinokabegakure-no-sato.local ‖ BitsCli ‖ 310 ‖ 13 ‖ - ‖ ErrorCode:2147942450 ‖ ../hayabusa-sample-evtx/YamatoSecurity/Persistence/T1197_BitsClient.evtx
```

File: src/timeline/search.rs [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+            all_field_info.as_str(),
+            evtx_file.as_str(),
+        ];
+        if wtr.as_ref().is_some() {
+            wtr.as_mut().unwrap().write_record(&record_data).ok();
+        } else {
+            wtr.as_mut()
+                .unwrap()
+                .write_field(record_data.join(" ‖ "))
+                .ok();
```
Comment (itiB) [inline src/timeline/search.rs:301]: @hitenkoku 
Omg, I just wanted to say thank you for implementing it quickly and easily, as I struggled to divide the search results with `||`.  Thank you:pray:

**Review by YamatoSecurity — APPROVED** [review]
@hitenkoku Looks great! Let's merge this and then the regex PR by @itiB

**Pull request merged**